pkg/apps/watch-appstream.py

```diff
@@ -0,0 +1,253 @@
+
+def element(xml, tag):
+    if lang:
+        for elt in xml.iter(tag):
+            if attr_lang(elt) == lang:
+                return elt
+    return xml.find(tag)
+
+def element_value(xml, tag):
+    elt = element(xml, tag)
+    return elt.text if elt is not None else None
+
+def convert_description(xml, use_lang = True):
+    if xml is None:
+        return None
+
+    want_lang = lang if use_lang else None
+
+    # Only the following constructs are allowed, and they all appear
+    # at the top level:
+    #
+    # <p>text</p>
+    # <ul><li>text</li>...</ul>
+    # <ol><li>text</li>...</ol>
+
+    # A description can have 'lang' attributes both on the actual
+    # <description> element and on the contained <p>, <ul>, and <ol>
+    # elements, but probably not on the <li> elements.
+
+    def text(xml):
+        return " ".join(xml.itertext())
+
+    res = [ ]
+    for c in xml:
+        if attr_lang(c) != want_lang:
+            continue
+        if c.tag == 'p':
+            res.append(text(c))
+        elif c.tag == 'ul' or c.tag == 'ol':
+            res.append({ "tag": c.tag, "items": map(text, list(c.findall('li'))) })
+
+    # If we found nothing that matches lang, fall back to default
+    if lang is not None and len(res) == 0:
+        res = convert_description(xml, False)
+
+    return res
+
+def convert_cached_icon(dir, origin, xml):
+    if 'width' in xml.attrib and 'height' in xml.attrib:
+        width = xml.attrib['width']
+        height = xml.attrib['height']
+    else:
+        width = "64"
+        height = "64"
+
+    return os.path.join(dir, "..", "icons", origin, width+"x"+height, xml.text)
+
+def convert_remote_icon(xml):
+    return xml.text
+
+def convert_local_icon(xml):
+    return xml.text
+
+def find_and_convert_icon(dir, origin, xml):
+    if xml is None:
+        return None
+
+    # We want the biggest one that isn't larger than 64x64
+    height = 0
+    icon = None
+    for ic in xml.iter('icon'):
+        h = int(ic.attrib['height'])
+        if h <= 64 and (h > height or icon is None):
+            icon = ic
+            height = h
+
+    if icon is not None:
+        if icon.attrib['type'] == "cached":
+            return convert_cached_icon(dir, origin, icon)
+        elif icon.attrib['type'] == "remote":
+            return convert_remote_icon(icon)
+        elif icon.attrib['type'] == "local":
+            return convert_local_icon(icon)
+
+    return None
+
+def convert_screenshots(xml):
+    if xml is None:
+        return [ ]
+
+    shots = [ ]
+    for sh in xml.iter('screenshot'):
+        for img in sh.iter('image'):
+            if img.attrib['type'] == "source":
+                shots.append({ 'full': img.text })
+
+    return shots
+
+def convert_launchables(xml):
+    ables = [ ]
+
+    for elt in xml.iter('launchable'):
+        type = elt.attrib['type']
+        if type == "cockpit-manifest":
+            ables.append({ "name": elt.text, "type": type })
+
+    return ables
```

Comment:
A-ha, I suppose that answers my error handling question from further above -- it could be that there is some XML which doesn't contain a cockpit-manifest, and thus there should be some user-visible error handling?

Reply:
I think I don't understand... If there is no `<launchable type"cockpit-manifest"/>`, the component will be filtered out completely.

Reply:
Right, as I wrote above I previously misunderstood this. It's fine.